### ISSUE TITLE
Allow configuration of irrelevant data truncation

### ISSUE THREAD

**Issue by jeaye** [bhb/expound#18]
The ommission of argument data, in failed calls to instrumented functions, is less helpful than clojure.spec would otherwise be. While I understand this brevity may be useful in some cases, my preference is to enjoy the formatting of expound without losing any data in translation. Are you open to making this truncation of irrelevant data configurable (say, using a dynamic var)?

Example:
```clojure
Call to #'flarb did not conform to spec:
<filename missing>:<line number missing>

Function arguments

-- Spec failed --------------------

  (... {:my-app.common.request/url-path "api/do-things",
        :my-app.common.request/method :my-app.common.request/post,
        :my-app.common.request/content-type :my-app.common.request/edn,
        :my-app.common.request/headers {"Authorization" "Bearer meow"},
        :my-app.common.request/body
        {:my-app.common.transit.challenge/id 1,
         :my-app.common.transit.challenge/num-games 0,
         :my-app.common.transit.money/money 1},
        :my-app.common.request/response-spec ::spam} ... ...)
       ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

should satisfy

  integer?



-------------------------
Detected 1 error
```

As you can see, it's hard to tell what's wrong here. This is partially due to the ommitted data and partially due to expound not showing the relevant specs in this case.

Thank you!

**Comment by bhb** [comment]
Thanks very much for reporting this. It looks like there are a few problems here. 

1. The relevant specs are not showing. Is there any information in `:via` in this case? Can you possibly share your specs?

2. It looks like there is a bug with underlining (with the "^" character) in this case. It should not be so wide given the data itself is more compact.

3. The way that printing of values works is not configurable. 

1 & 2 look like potential bugs to me, so I can fix them in the short term.

3 is something I'm very open to, but I need to give it a little more thought. Dynamic variables is one way to go, and the nice thing about dynamic vars is that it is easy to make minor adjustments (like not omitting data). OTOH, it might be more generally useful to refactor Expound so that users can easily build their own `printer` function using some helper functions. Then again, maybe this is too much effort for basic configuration. I'll think more about it.

**Comment by bhb** [comment]
Short discussion of goals on slack https://clojurians-log.clojureverse.org/clojure-spec/2017-07-17.html#inst-2017-07-17T18:09:19.667294Z

**Comment by bhb** [comment]
The other thing that might (potentially) help here is if I could show the actually calling with the function and arg names. I'm not sure if that's possible though - something to look into.

**Comment by jeaye** [comment]
Ok, so, there's some confusion here, caused by me. I'll address your points each in turn.

1. I have tried testing this locally, in Clojure and ClojureScript, and have not reproduced the issue. The issue was reported to me by a teammate, but it seems like he wasn't on the latest expound/orchestra, so the `:via` wasn't available. IFF this shows up again and I can get a good repro case for it shall I then bring it back up. Sorry for the confusion.

2. My fault again; the output was much longer, since it contained an authentication token which I snipped into `meow`. The underlining has been working just fine for me.

3. This looks like it's the only thing we need to discuss at this point.

Thanks for the thoughtful replies and sorry again for the misdirection.

**Comment by bhb** [comment]
@jeaye Thanks very much for the clarification! I appreciate the follow up.

I've been chasing down some unrelated bugs, but I'm still thinking about this. Another third option occurred to me: I could provide a new printer named something like `instrumentation-printer` that is optimized for instrumentation (so far, that only means not omitting values, but perhaps there are other changes I might make). I think that might be overkill for this case, but custom printers have been mentioned before (for instance, to make a public-facing error messages for an API), so it's something I'm keeping in mind.

**Comment by bhb** [comment]
@jeaye I'm thinking more about how we'd use this. If this was configurable, do you expect that you would enable it globally for your project? Or using `binding` to manually set it when the default printer failed to help locate the error?

**Comment by jeaye** [comment]
Whichever approach you choose: dynamic var, custom printer, or something else, my team will be using it globally in both our client and server. Orchestra runs globally and we want good spec errors globally, too!

**Comment by bhb** [comment]
Excellent. Thanks for the information!

**Comment by bhb** [comment]
@jeaye Can you try out version 0.2.2-SNAPSHOT? You can follow [these instructions](https://github.com/bhb/expound/blob/f2e967ed29789e424ea8561b07303cd2279cad7f/README.md#configuring-the-printer) to set up a custom printer. Please let me know if it works for your project. 

Thanks for suggesting this change!

**Comment by jeaye** [comment]
Most excellent! I'll give this a shot tomorrow. Thanks for the quick work and dependable documentation.

**Comment by jeaye** [comment]
Trying this now. Sorry for the delay!

**Comment by jeaye** [comment]
Yeah, this `{:show-valid-values? true}` does the trick. Thanks so much. It also raises another problem, however, which imposes another request for configuration onto expound. With valid values not being elided, the spec explanations can be so long that Android's logging system cuts them off (it only allows like 10 lines or so).

My first thought was "No problem, I'll use a custom printer to do a `with-out-str` on the default printer and then chunk the logs out to Android so I can see everything." Unfortunately, the current printing customization only allows for the specification of a pure `:value-str-fn`. Perhaps I could do the logging in there and return an empty string, but I haven't given that a shot yet. Suggestions?

**Comment by jeaye** [comment]
Oh, to follow up, the `:value-str-fn` only covers the rendering of the value. So it looks like I could override `s/*explain-out*` myself, have it call `expound/custom-printer` within a `with-out-str` and then chunk that.

**Comment by jeaye** [comment]
I'm closing this, since the snapshot works exactly as expected. Thanks, again, for the help and I'm looking forward to a full `0.2.2` release. :)

**Comment by bhb** [comment]
@jeaye Thanks very much for reporting this and for testing the fix!

**Comment by bhb** [comment]
I've released [Expound 0.3.0](https://github.com/bhb/expound/blob/master/CHANGELOG.md#030---2017-09-05)

**Comment by jeaye** [comment]
Alright! Awesome work.
